### ISSUE TITLE
Figure out our data/CRM solution

### ISSUE THREAD

**Issue by saikat** [BrandNewCongress/welcome#28]
## The Problem
We are currently using a bunch of different tools for our data, which leads to us having our data fragmented out and makes it difficult to get an accurate picture of our supporters at an individual level and also to do proper targeting of our supporters.  Currently, we are using Airtable to track our candidate nomination pipeline, Nationbuilder (soon to be Blue State Digital) to track our full list of supporters and volunteers, Airtable to track which volunteers are on what team, a new events system to track who has hosted an event or attended an event, and, in the future, we plan to have an auto-dialer (which will then be the store of who has made how many calls), a texting tool (which will be another store of work people have done), etc.  

But this is not simply a problem that Brand New Congress has -- every campaign deals with the issue of having data in multiple silos.  The traditional way a campaign gets set up is it will have one tool that it uses to track its e-mail list of supporters (Nationbuilder, Blue State Digital, etc.), and then usually will set up a second tool to handle canvassing and field operations (NGP VAN), and then have a multitude of Google Sheets that track all kinds of other custom data models/volunteer work related data, and then other tools like an auto-dialer or texting tool that also have silos of their own data.  Then, to actually try to get a view across all these data platforms, the campaign will usually hack together scripts to try to shuttle data from these other tools into some central platform (either VAN or Nationbuilder/Blue State Digital or, in the case of a Democratic campaign, there's often a big central database called Vertica that everything syncs into) creating a mess of jobs and usually leaving all the random data in Google Sheets behind because it's simply not worth trying to capture everything.

You might ask "why not just try to use Nationbuilder or Blue State Digital for everything?" and the basic answer is that these tools are simply not the best tool out there for doing everything.  Blue State Digital, for example, has a good mailer but pales in comparison to VAN for tools to do field work and 'cut turf' for canvassing.  Similarly, VAN has a phone calling tool, but it's not an auto-dialer, and so is no where near as efficient as an auto-dialer.  And the plethora of Google Sheets people create also show a real need for people needing one-off data models to handle certain workflows or volunteer flows.  And, on top of this, new tools are constantly being built that do parts of the work of a campaign better than previous.  So while a tool like VAN or BSD are built to be these monolithic, all-in-one solutions, the truth is they are not the best tool for all the jobs so we end up using various other software with their own data models that then end up being a pain to sync with.

You might also ask "why do we actually need to sync all this data -- isn't it fine to have this data in different places?" and the answer there is twofold: 1) we need to be able to query across all this data to do proper targeting and 2) it is incredibly useful to be able to see the full picture of a certain supporter or volunteer.  2 is simpler to explain -- basically when I look at a volunteer, I'd love to be able to see how many phone calls she's made, what volunteer teams she's been on, etc. to get a sense of what this person is actually good and and where she might fit in the organization.  For 1 -- a simple example is wanting to send an e-mail to all our event hosts who live in New York who have made more than 100 phone calls.  If event data is in one tool and phone call data is in another tool and where the person lives is in Nationbuilder, doing that kind of targeting becomes really difficult.

There is also another problem with the current landscape of campaign CRMs (things like NGP VAN, BSD, Nationbuilder, Action Network, etc.) -- none of them work very well for modern, volunteer-driven campaigns.  What I mean by this is that all these tools have onerous, whitelist based permission models whereas a lot of the way modern day campaigns work is that there are a huge pool of volunteers who need access to portions of the data to use to do work.  Here's an example use case -- when supporters create events, we want to be able to approve those events.  The traditional way to do this in, say, Blue State Digital would be to give one of your campaign staff event moderation privileges in BSD.  That person would then log in, be able to see all the events and event RSVPs, and go through and approve them one-by-one.  But the way a big volunteer-driven campaign like Bernie's campaign or Brand New Congress would do this is create a volunteer team of event moderators who each get assigned a number of events to approve or reject.  This doesn't work with BSD because simply adding all these volunteers to BSD means a) they step all over each other's toes in the event moderation queue and b) every volunteer ends up getting access to basically all the data.  Additionally, volunteers can be fluid -- someone with time this week may not have time next week -- so dealing with constantly adding and unadding people to BSD becomes its own headache.  Instead, what this boils down to, is being able to give volunteers work to do in a very simple way (e.g. with a link) that gives them access to just the amount of data they need to do the work.

## The Solution?

The problems listed call for a modern day campaign CRM that acknowledges that the ecosystem of tools for campaigning need to live outside the campaign database.  I'm imagining a very lightweight central data warehouse of all the campaign data that has only a few components:
1. It will have one model for sure -- a People model
1. It will have the ability to extend the People model and attach related data to it.  E.g. data from a voter file could be attached to this model
1. It will have a GraphQL API with complete filtering capabilities across all the models in the database. 
1. The GraphQL API will have the capability of saving filters (either as static or streaming -- streaming means it's a dynamic list that continues to update as new data enters into it) at unique random URLs.
1. It will have the ability to create related models to the people via scheduled jobs that are linked to outside APIs to fetch and sync data into the database.  E.g. you could integrate data from your autodialer by setting up a scheduled job that would fetch data from the autodialer API and pull it in to the database.

Then, in addition to this, we'd want to build a UI (right now I'm thinking something that looks a lot like Airtable) on top of the GraphQL API that lets us both do filtering via this UI and also create saved filters.

This would be the extent of what exists in this tool.  And all of the common CRM tools -- like a mailer, dialer, etc. -- would all be standalone tools that live outside of this and interact with it via the GraphQL API.  So targeting would be done by creating a new filter and then a tool could subscribe to that filter (if it wants streaming updates) or just fetch the data from the filter once.

And similarly, we could use these filters to give volunteers access to subsets of the data.  So you could imagine a filter saved at https://api.brandnewcongres.org/searches/sntahoeurlf234 would, when simply fetched, fetch the data as a JSON, but when fetched within the UI, show the data in grid format.  And each filter would, in addition, have permissions set so that the UI could allow editing the data or not. I haven't really thought through if this is actually useful (vs. the tools accessing the data being the place to cut lists for volunteers), but it seems possibly useful.

So basically, this boils down to:

1. Create the database that will store all this
1. Create the jobs that allow this database to sync in data from other sources
1. Create the GraphQL API to access this data
1. Create a grid-based UI to interact with the data

Also, listed below are some tools that have served as inspiration for a lot of this and possibly could be useful when building this out.
* Graph.cool
* Scaphold
* Airtable
* Knack
* Adminium
* Sonadier
* Fieldbook
* Baqend
* Firebase
